design/control-data-plane-separation/design.md

```diff
@@ -0,0 +1,770 @@
+command and then download the config chunks over the `DownloadChannel .`It will then assemble the chunks into a full
+configuration and attempt to apply it. It will send the status of the configuration application over
+the `CommandChannel` in a `NginxConfigResponse` message.
+
+[internal-config]: https://github.com/nginxinc/nginx-kubernetes-gateway/blob/main/internal/nginx/config/http/config.go
+
+[zip-file]: https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/proto/common.pb.go#L668
+
+[zip-writer]: https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/zip/zipped_file.go#L36
+
+[directory-map]: https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/proto/common.pb.go#L99
+
+[action]: https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/proto/nginx.pb.go#L26
+
+[config-descriptor]:https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/proto/config.pb.go#L81
+
+[gw-tls-config]: https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1beta1.GatewayTLSConfig
+
+[sds]: https://www.envoyproxy.io/docs/envoy/latest/configuration/security/secret
+
+[host-path]: https://kubernetes.io/docs/concepts/storage/volumes/#hostpath
+
+[config-resp]: https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/proto/command.pb.go#L891
+
+## Agent Configuration
+
+We can configure the agent through a YAML file, command-line flags, or environment variables on start-up. The agent
+interprets configuration in the following priorities (from highest to lowest):
+
+1. Command-line flags
+2. Environment variables
+3. Config files
+
+We can either configure the agent with command-line flags or mount the config file as a volume to the agent’s container.
+Some configuration options we care about are:
+
+- `server-host`: the DNS name of the control plane Kubernetes Service
+- `server-grpcport`: the gRPC port that the control plane is listening on
+- `tls-ca`: the path to the CA certificate file
+- `tls-cert`: the path to the agent’s certificate file
+- `tls-key`: the path to the agent’s private key file
+- `tls-enable`: enable TLS for secure communication with the control plane
+- `features`: list of enabled features
+- `config-dirs`:the paths that you want to grant agent read/write access to
+
+For a list of configuration options, see [NGINX Agent CLI Flags & Usage][cli].
+
+The control plane can send runtime configuration using the [`AgentConfig`][agent-config] messages. I’m not sure if we
+have a use case for runtime configuration at the moment.
+
+[agent-config]: https://github.com/nginx/agent/blob/ea3a1b4df5d7ecf95bd3d9297d26e420f5e1dd57/sdk/proto/agent.pb.go#L320
+
+[cli]: https://docs.nginx.com/nginx-management-suite/nginx-agent/install-nginx-agent/#nginx-agent-cli-flags-usage
+
+## Data Plane Scaling
```

Comment:
Do we need to consider control plane upgrades? And also data plane (with agent) upgrades?
with the goal of making zero-downtime upgrades -- no client traffic is lost

One problem I see is potentially the agent making a breaking change that requires the use of the new control plane version (making the change to the control plane)